lib/util/cmcd_batching_manager.js

```diff
@@ -0,0 +1,296 @@
+ */
+
+goog.provide('shaka.util.CmcdBatchingManager');
+goog.require('shaka.log');
+goog.require('shaka.net.NetworkingEngine');
+goog.require('shaka.util.Timer');
+goog.require('shaka.util.CmcdUtils');
+goog.require('shaka.util.StringUtils');
+goog.requireType('shaka.Player');
+
+/**
+ * @summary
+ * A CmcdBatchingManager handles batching of CMCD reports for efficient
+ * delivery to collection endpoints.
+ */
+shaka.util.CmcdBatchingManager = class {
+  /**
+   * @param {shaka.Player} player
+   */
+  constructor(player) {
+    /** @private {shaka.Player} */
+    this.player_ = player;
+
+    /** @private {!Map<string, shaka.util.CmcdBatchingManager.Batch>} */
+    this.batches_ = new Map();
+
+    /** @private {!Map<string, !shaka.util.Timer>} */
+    this.timers_ = new Map();
+
+    /** @private {!Array<shaka.util.CmcdBatchingManager.RetryItem>} */
+    this.retryQueue_ = [];
+
+    /** @private {!Map<string, !shaka.util.Timer>} */
+    this.retryTimers_ = new Map();
+
+    /** @private {!Array<number>} */
+    this.retryDelays_ = [100, 500, 1000, 3000, 5000]; // ms
+
+    /** @private {!Set<string>} */
+    this.goneUrls_ = new Set();
+  }
+
+  /**
+   * Add a CMCD report to be batched and sent to the specified target.
+   *
+   * @param {!shaka.extern.CmcdTarget} target The CMCD target configuration
+   * @param {!CmcdData} cmcdData The CMCD data to batch
+   * @suppress {checkDebuggerStatement}
+   */
+  addReport(target, cmcdData) {
+    if (this.goneUrls_.has(target.url)) {
+      return;
+    }
+
+    const key = this.generateTargetKey_(target);
+    const serializedCmcd = shaka.util.CmcdUtils.serialize(cmcdData);
+    const urlEncoded = encodeURIComponent(serializedCmcd);
+
+    if (!this.batches_.has(key)) {
+      this.batches_.set(key, {
+        cmcdData: '',
+        target: target,
+      });
+    }
+
+    const batch = this.batches_.get(key);
+    if (batch.cmcdData === '') {
+      batch.cmcdData = urlEncoded;
+    } else {
+      batch.cmcdData += '\n' + urlEncoded;
+    }
+
+    // Set up batch timer if configured and not already set
+    if (target.batchTimer && !this.timers_.has(key)) {
+      const timer = new shaka.util.Timer(() => {
+        this.flushByTargetKey_(key);
+      });
+      timer.tickEvery(target.batchTimer);
+      this.timers_.set(key, timer);
+    }
+
+    // Check if batch size limit is reached
+    if (target.batchSize &&
+        batch.cmcdData.split('\n').length >= target.batchSize) {
+      this.flushByTargetKey_(key);
+    }
+  }
+
+  /**
+   * Flush all batches for a specific URL.
+   *
+   * @param {string} url The URL to flush batches for
+   */
+  flushBatch(url) {
+    for (const key of this.batches_.keys()) {
+      const batch = this.batches_.get(key);
+      if (batch && batch.target && batch.target.url === url) {
+        this.flushByTargetKey_(key);
+      }
+    }
+  }
+
+  /**
+   * Reset the batching manager, clearing all batches and timers.
+   */
+  reset() {
```

Comment:
Maybe call it `release()` and make this class to implement `IReleasable` interface? That would be more easy to follow by devs who work on shaka day-by-day.